### PR TITLE
CAPE

### PR DESCRIPTION
**Pull request by tojiboyevf**

### CODE REVIEW

File: CAPE/CAPE/run_cape_offline.cpp [hunk 1]
```diff
@@ -97,19 +95,25 @@ void organizePointCloudByCell(Eigen::MatrixXf & cloud_in, Eigen::MatrixXf & clou
...
-        sequence = "cy_0";
+        string_buff<<"input/"<<argv[2];
+    } else {
+        PATCH_SIZE = 16;
+        string_buff<<"input";
+    }
+    
+    for (int i = 1; i < argc; ++i) {
+        if (std::string(argv[i]) == "--vis") {
+            show_visualization=true;
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp]: spaces

File: CAPE/CAPE/run_cape_offline.cpp [hunk 2]
```diff
@@ -97,19 +95,25 @@ void organizePointCloudByCell(Eigen::MatrixXf & cloud_in, Eigen::MatrixXf & clou
 
 int main(int argc, char ** argv){
 
-    string sequence;
+    bool show_visualization=false;
+    stringstream string_buff;
+
     int PATCH_SIZE;
-    if (argc>1){
+    if (argc>2){
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp:107]: What is the command parsed here? Nothing about it in README?

File: CAPE/CAPE/run_cape_offline.cpp [hunk 3]
```diff
@@ -197,20 +199,26 @@ int main(int argc, char ** argv){
...
+    int frame_num=0;
+    DIR *dir;
+    struct dirent *ent;
+    if ((dir = opendir (string_buff.str().c_str())) != NULL) {
+    while ((ent = readdir (dir)) != NULL) {
+        if(boost::algorithm::contains(ent->d_name, ".png")) frame_num++;
+    }
+    closedir (dir);
+    } else {
+    perror ("could not open directory");
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp]: Tabs, spaces (codestyle)

**Review by DmiitriyJarosh — CHANGES_REQUESTED** [review]

File: CAPE/CAPE/run_cape_offline.cpp (hunk 2 above)
Comment (tojiboyevf) [reply CAPE/CAPE/run_cape_offline.cpp:107]: Here I did visualization optionally. You can change pixel size or you can show another folder that contains depth images. I added this information in README but then you told me that if we work with ENTRYPOINT then no needed part about parameters and so on. That's why I deleted it from README

Comment (tojiboyevf) [reply CAPE/CAPE/run_cape_offline.cpp:107]: Now I added these parameters with examples in docker run command, but I think if I showed it without docker run command it might be more understandable.

File: CAPE/CAPE/run_cape_offline.cpp [hunk 4]
```diff
@@ -1,15 +1,13 @@
-/*
- * Copyright 2018 Pedro Proenza <p.proenca@surrey.ac.uk> (University of Surrey)
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp]: Why did you remove copyright?

File: CAPE/CAPE/run_cape_offline.cpp [hunk 5]
```diff
@@ -1,15 +1,13 @@
...
- */
-
 #include <iostream>
 #include <cstdio>
 #define _USE_MATH_DEFINES
 #include <math.h>
 #include <opencv2/opencv.hpp>
 #include <Eigen/Dense>
 #include "CAPE.h"
+#include<string>
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp]: space is skipped `#include<string>` -> `#include <string>`

**Review by DmiitriyJarosh — CHANGES_REQUESTED** [review]

File: CAPE/CAPE/run_cape_offline.cpp [hunk 6]
```diff
@@ -97,20 +100,22 @@ void organizePointCloudByCell(Eigen::MatrixXf & cloud_in, Eigen::MatrixXf & clou
 
 int main(int argc, char ** argv){
 
-    string sequence;
+    bool show_visualization= false;
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp]: space `bool show_visualization= false;` -> `bool show_visualization = false;`

File: CAPE/CAPE/run_cape_offline.cpp [hunk 7]
```diff
@@ -97,20 +100,22 @@ void organizePointCloudByCell(Eigen::MatrixXf & cloud_in, Eigen::MatrixXf & clou
...
+
     int PATCH_SIZE;
-    if (argc>1){
+    if (argc>2){
         PATCH_SIZE = atoi(argv[1]);
-        sequence = argv[2];
-    }else{
-        PATCH_SIZE = 20;
-        sequence = "cy_0";
+        string_buff<<"input/"<<argv[2];
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp]: `<<` is also an operator -- so use spaces))  `string_buff << "input/" << argv[2];` -> `string_buff << "input/" << argv[2];`

File: CAPE/CAPE/run_cape_offline.cpp [hunk 8]
```diff
@@ -97,20 +100,22 @@ void organizePointCloudByCell(Eigen::MatrixXf & cloud_in, Eigen::MatrixXf & clou
...
+    }else {
+        PATCH_SIZE = 16;
+        string_buff<<"input";
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp]: spaces too

File: CAPE/CAPE/run_cape_offline.cpp [hunk 9]
```diff
@@ -197,20 +196,26 @@ int main(int argc, char ** argv){
     color_code[52][0] = 0; color_code[52][1] = 255; color_code[52][2] = 51;
     color_code[53][0] = 153; color_code[53][1] = 0; color_code[53][2] = 255;
 
+    int frame_num = 0;
+    DIR *dir;
+    struct dirent *ent;
+    if ((dir = opendir (string_buff.str().c_str())) != NULL) {
+    while ((ent = readdir (dir)) != NULL) {
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp]: tabs are incorrect

File: CAPE/CAPE/run_cape_offline.cpp [hunk 10]
```diff
@@ -197,20 +196,26 @@ int main(int argc, char ** argv){
...
+        return EXIT_FAILURE;
+    }
+
     // Initialize CAPE
     plane_detector = new CAPE(height, width, PATCH_SIZE, PATCH_SIZE, cylinder_detection, COS_ANGLE_MAX, MAX_MERGE_DIST);
 
-    int i=0;
-    while(true){
+    int i = 0;
+    while(i<frame_num){
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp]: spaces `while(i<frame_num){` -> `while (i < frame_num) {`

File: CAPE/CAPE/run_cape_offline.cpp [hunk 11]
```diff
@@ -301,8 +296,16 @@ int main(int argc, char ** argv){
                 cv::rectangle(seg_rz,  cv::Point(width/2 + 80+15*j,6),cv::Point(width/2 + 90+15*j,16), cv::Scalar(color_code[cylinder_code_offset+j][0],color_code[cylinder_code_offset+j][1],color_code[cylinder_code_offset+j][2]),-1);
             }
         }
-        cv::imshow("Seg", seg_rz);
-        cv::waitKey(1);
+
+        save_path.str("");
+        save_path<<"output/segment_"<<i<<".png";
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp]: spaces

File: CAPE/CAPE/run_cape_offline.cpp [hunk 12]
```diff
@@ -301,8 +296,16 @@ int main(int argc, char ** argv){
...
+        cv::imwrite(save_path.str(), seg_rz);
+        if(show_visualization){
```
Comment (DmiitriyJarosh) [inline CAPE/CAPE/run_cape_offline.cpp]: spaces `if(show_visualization){` -> `if (show_visualization) {`

**Review by DmiitriyJarosh — CHANGES_REQUESTED** [review]

**Review by DmiitriyJarosh — APPROVED** [review]

**Review by anastasiia-kornilova — APPROVED** [review]
